[DOCS-3095] - local mongo collection (#3673)
viamrobotics/docs: docs/services/data/_index.md

server and write captured tabular data to the configured `mongo_capture_config.database` and `mongo_capture_config.collection`
(or their defaults if unconfigured) after enqueuing that data to be written to disk.

If writes to MongoDB fail for any resson, Data Capture will log an error for each failed write and continue capturing.

Failing to write to MongoDB doesn't affect capturing & syncing data to app.viam.com other than adding capture latency.

{{< alert title="Caution" color="caution" >}}

- Capturing Directly to MongoDB may write data to MongoDB that later fails to be written to disk (and therefore never gets synced to app.viam.com).
- Capturing Directly to MongoDB does not retry failed writes to MongoDB. As a consequence, it is NOT guaranteed all data captured will be written to MongoDB. This can happen in cases such MongoDB being inaccessible to viam-server or writes timing out.
- Capturing Directly to MongoDB may reduce the maximum frequency that Data Capture can capture data due to the added latency of writing to MongoDB. If your usecase needs to support very high capture rates, this feature may not be appropriate.

> {{< /alert >}}

Author: I'm confused why this is showing up in red with a title of `WARNING` rather than in yellow with a title of `Caution`:
 
![Screenshot 2024-11-13 at 12 40 09 PM](https://github.com/user-attachments/assets/75b45295-a250-4e59-89ae-18f034906964)

Author: It appears as if the demo site is not updating,

This is what is looks like locally when I run `make serve-dev`
![Screenshot 2024-11-13 at 1 16 57 PM](https://github.com/user-attachments/assets/972bd6f0-8d58-43f1-97e7-7d6dcd128785)

Author: it's now working now that I rebased & squashed

{{< /expand >}}

{{< expand "Capture directly to MongoDB" >}}

Data capture supports capturing tabular data directly to MongoDB in addition to capturing to disk.

This feature is intended to support use cases like offline dashboards which don't require strong data delivery or consistency guarantees.

> Here is a sample configuration that will capture fake sensor readings both to the configured MongoDB URI as well as to the Viam capture directory on disk:

Reviewer: ```suggestion
Here is a sample configuration that will capture fake sensor readings both to the configured MongoDB URI as well as to the `~/.viam/capture` directory on disk:
```

{{< /expand >}}

{{< expand "Capture Directly to MongoDB" >}}

> Data Capture supports capturing tabular data directly to MongoDB in addition to capturing to disk.

Reviewer 2: ```suggestion
The data management service supports capturing tabular data directly to MongoDB in addition to capturing to disk.
```

Author: https://github.com/viamrobotics/docs/pull/3673#discussion_r1842392926

> This feature is intended to support usecases like offline dashboards which don't require strong data

Reviewer 2: nit? https://github.com/search?q=repo%3Aviamrobotics%2Fdocs+%22use+case%22&type=code

```suggestion
This feature is intended to support use cases like offline dashboards which don't require strong data
```

delivery or consistency guarantees.

> Here is a sample configuration that will capture fake sensor readings both to the configured MongoDB URI

Reviewer: ```suggestion
Here is a sample configuration that will capture fake sensor readings both to the configured MongoDB URI as well as to the `~/.viam/capture` directory on disk:
```

> Here is a sample cloud config that will capture fake sensor readings both to the configured MongoDB URI
> as well as to the viam capture directory on disk:

Reviewer 2: ```suggestion
Here is a sample configuration that will capture fake sensor readings both to the configured MongoDB URI
as well as to the Viam capture directory on disk:
```

```json
{
  "components": [
    {
      "name": "sensor-1",
      "namespace": "rdk",
      "type": "sensor",
      "model": "fake",
      "attributes": {},
      "service_configs": [
        {
          "type": "data_manager",
          "attributes": {
            "capture_methods": [
              {
                "method": "Readings",
                "capture_frequency_hz": 0.5,
                "additional_params": {}
              }
            ]
          }
        }
      ]
    }
  ],
  "services": [
    {
      "name": "data_manager-1",
      "namespace": "rdk",
      "type": "data_manager",
      "attributes": {
        "mongo_capture_config": {
          "uri": "mongodb://127.0.0.1:27017/?directConnection=true&serverSelectionTimeoutMS=2000"
        }
      }
    }
  ]
}
```

> When `mongo_capture_config.uri` is configured, Data Capture will attempt to connect to the configured MongoDB

Author: https://github.com/viamrobotics/docs/pull/3673#discussion_r1842392926

server and write captured tabular data to the configured `mongo_capture_config.database` and `mongo_capture_config.collection`
(or their defaults if unconfigured) after enqueuing that data to be written to disk.

> If writes to MongoDB fail for any resson, Data Capture will log an error for each failed write and continue capturing.

Author: https://github.com/viamrobotics/docs/pull/3673#discussion_r1842392926

> If writes to MongoDB fail for any resson, Data Capture will log an error for each failed write and continue capturing.

Reviewer 2: ok but notice the typo `s/resson/reason/`

> If writes to MongoDB fail for any resson, Data Capture will log an error for each failed write and continue capturing.

Author: thanks! Fixed

Failing to write to MongoDB doesn't affect capturing & syncing data to app.viam.com other than adding capture latency.

{{< alert title="Caution" color="caution" >}}

- Capturing Directly to MongoDB may write data to MongoDB that later fails to be written to disk (and therefore never gets synced to app.viam.com).
- Capturing Directly to MongoDB does not retry failed writes to MongoDB. As a consequence, it is NOT guaranteed all data captured will be written to MongoDB. This can happen in cases such MongoDB being inaccessible to viam-server or writes timing out.

> - Capturing Directly to MongoDB may reduce the maximum frequency that Data Capture can capture data due to the added latency of writing to MongoDB. If your usecase needs to support very high capture rates, this feature may not be appropriate.

Author: I've taken the parts of this suggestion which don't pertain to removing reference to `Data Capture` due to https://github.com/viamrobotics/docs/pull/3673#discussion_r1842392926

| `maximum_num_sync_threads` | int | Optional | Max number of CPU threads to use for syncing data to the Viam Cloud. <br> Default: [runtime.NumCPU](https://pkg.go.dev/runtime#NumCPU)/2 so half the number of logical CPUs available to viam-server |  <p class="center-text"><i class="fas fa-times" title="no"></i></p> |
| `mongo_capture_config.uri` | string | Optional | The [MongoDB URI](https://www.mongodb.com/docs/v6.2/reference/connection-string/) data capture will attempt to write tabular data to after it is enqueued to be written to disk. When non empty, data capture will capture tabular data to the configured MongoDB database & collection at that URI.<br>See  `mongo_capture_config.database` and  `mongo_capture_config.collection` below for database & collection defaults. |  <p class="center-text"><i class="fas fa-times" title="no"></i></p> |
| `mongo_capture_config.database` | string | Optional | When `mongo_capture_config.uri` is non empty, changes the database data capture will write tabular data to. <br> Default: `"sensorData"`   |  <p class="center-text"><i class="fas fa-times" title="no"></i></p> |

> | `mongo_capture_config.collection` | string | Optional | When `mongo_capture_config.uri` is non empty, changes the collection data capture will write tabular data to.<br> Default: `"readings"`   |  <p class="center-text"><i class="fas fa-times" title="no"></i></p> |

Author: Question for the docs team: Is this the proper way to document a config attribute which is an object with multiple fields? I wasn't sure how best to document each field of the `mongo_capture_config` object.

| `sync_interval_mins` | float | Optional | Time interval in minutes between syncing to the cloud. Viam does not impose a minimum or maximum on the frequency of data syncing. However, in practice, your hardware or network speed may impose limits on the frequency of data syncing. <br> Default: `0.1`, meaning once every 6 seconds. |  <p class="center-text"><i class="fas fa-check" title="yes"></i></p> |
| `delete_every_nth_when_disk_full` | int | Optional | How many files to delete when local storage meets the [fullness criteria](/services/data/#storage). The data management service will delete every Nth file that has been captured upon reaching this threshold. Use JSON mode to configure this attribute. <br> Default: `5`, meaning that every fifth captured file will be deleted. |  <p class="center-text"><i class="fas fa-times" title="no"></i></p> |
| `maximum_num_sync_threads` | int | Optional | Max number of CPU threads to use for syncing data to the Viam Cloud. <br> Default: [runtime.NumCPU](https://pkg.go.dev/runtime#NumCPU)/2 so half the number of logical CPUs available to viam-server |  <p class="center-text"><i class="fas fa-times" title="no"></i></p> |

> | `mongo_capture_config.uri` | string | Optional | The [MongoDB URI](https://www.mongodb.com/docs/v6.2/reference/connection-string/) data capture will attempt to write tabular data to after it is enqueued to be written to disk. When non empty, data capture will capture tabular data to the configured MongoDB database & collection at that URI.<br>See  `mongo_capture_config.database` and  `mongo_capture_config.collection` below for database & collection defaults.<br>See [Data Capture Directly To MongoDB](/services/data/#capture-directly-to-mongodb) for an example config.|  <p class="center-text"><i class="fas fa-times" title="no"></i></p> |

Reviewer 2: nit - 'data capture' is not a service from the user's perspective I think?

```suggestion
| `mongo_capture_config.uri` | string | Optional | The [MongoDB URI](https://www.mongodb.com/docs/v6.2/reference/connection-string/) that the data management service will attempt to write tabular data to after it is enqueued to be written to disk. When non empty, the data management service will capture tabular data to the configured MongoDB database & collection at that URI.<br>See  `mongo_capture_config.database` and  `mongo_capture_config.collection` below for database & collection defaults.<br>See [Data Capture Directly To MongoDB](/services/data/#capture-directly-to-mongodb) for an example config.|  <p class="center-text"><i class="fas fa-times" title="no"></i></p> |
```

Author: Data Capture is not a service but it is a feature of Data Manager. I think this is indeed in line with the existing nomenclature of these docs.

See at the top of the docs:
![Screenshot 2024-11-14 at 10 09 10 AM](https://github.com/user-attachments/assets/32047db9-cfe2-4c86-9618-df8f6109590a)

I think it is important to document that this is a feature of Data Capture as users need to understand that if they disable data capture they will stop writing data to mongo, where as if they are capturing & not syncing, they will still write data to mongo.

Reviewer 2: ok -- I was worried of referring to Data Capture (capital D capital C) for the first time in the doc. in your screenshot, the word 'capture' is used for its functional meaning, not as a name of a _thing_. I'll yield to your judgement and have approved the change already, just wanted to note this difference.

Author: I'll downcase data capture to keep it consistent with the rest of the docs.

Reviewer 2: oh yeah that works -- thanks! :)

> | `mongo_capture_config.database` | string | Optional | When `mongo_capture_config.uri` is non empty, changes the database data capture will write tabular data to. <br> Default: `"sensorData"`   |  <p class="center-text"><i class="fas fa-times" title="no"></i></p> |
> | `mongo_capture_config.collection` | string | Optional | When `mongo_capture_config.uri` is non empty, changes the collection data capture will write tabular data to.<br> Default: `"readings"`   |  <p class="center-text"><i class="fas fa-times" title="no"></i></p> |

Reviewer 2: again

```suggestion
| `mongo_capture_config.database` | string | Optional | When `mongo_capture_config.uri` is non empty, changes the database that the data management service will write tabular data to. <br> Default: `"sensorData".`   |  <p class="center-text"><i class="fas fa-times" title="no"></i></p> |
| `mongo_capture_config.collection` | string | Optional | When `mongo_capture_config.uri` is non empty, changes the collection that the data management service will write tabular data to.<br> Default: `"readings".`   |  <p class="center-text"><i class="fas fa-times" title="no"></i></p> |
```
